lib/PrintAsClang/ClangSyntaxPrinter.cpp

```diff
@@ -185,12 +187,92 @@ void ClangSyntaxPrinter::printValueWitnessTableAccessSequenceFromTypeMetadata(
 }
 
 void ClangSyntaxPrinter::printCTypeMetadataTypeFunction(
-    const NominalTypeDecl *typeDecl, StringRef typeMetadataFuncName) {
+    const NominalTypeDecl *typeDecl, StringRef typeMetadataFuncName,
+    llvm::ArrayRef<GenericRequirement> genericRequirements) {
+  // FIXME: Support generic requirements > 3.
+  if (!genericRequirements.empty())
+    os << "static_assert(" << genericRequirements.size()
+       << " <= " << NumDirectGenericTypeMetadataAccessFunctionArgs
+       << ", \"unsupported generic requirement list for metadata func\");\n";
   os << "// Type metadata accessor for " << typeDecl->getNameStr() << "\n";
   os << "SWIFT_EXTERN ";
   printSwiftImplQualifier();
   os << "MetadataResponseTy " << typeMetadataFuncName << '(';
   printSwiftImplQualifier();
-  os << "MetadataRequestTy)";
+  os << "MetadataRequestTy";
+  if (!genericRequirements.empty())
+    os << ", ";
+  llvm::interleaveComma(genericRequirements, os,
+                        [&](const GenericRequirement &) {
+                          // FIXME: Print parameter name.
+                          os << "void * _Nonnull";
+                        });
+  os << ')';
   os << " SWIFT_NOEXCEPT SWIFT_CALL;\n\n";
 }
+
+void ClangSyntaxPrinter::printGenericTypeParamTypeName(
+    const GenericTypeParamType *gtpt) {
+  os << "T_" << gtpt->getDepth() << '_' << gtpt->getIndex();
+}
+
+void ClangSyntaxPrinter::printGenericSignature(
+    const CanGenericSignature &signature) {
+  os << "template<";
+  llvm::interleaveComma(
+      signature.getGenericParams(), os,
+      [&](const CanTypeWrapper<GenericTypeParamType> &genericParamType) {
+        os << "class ";
+        printGenericTypeParamTypeName(genericParamType);
+      });
+  os << ">\n";
+  os << "requires ";
```

Comment:
Sorry if I missed this: does the generated header always require C++20?

Reply:
At the moment yes, this will be relaxed in the future.